File: bot-api/dotnet/test/src/BaseBotConstructorPrecedenceTest.cs
using System;
using NUnit.Framework;
using Robocode.TankRoyale.BotApi.Tests.Test_utils;
using static Robocode.TankRoyale.BotApi.Tests.Test_utils.EnvironmentVariables;

namespace Robocode.TankRoyale.BotApi.Tests;

[TestFixture]
public class BaseBotConstructorPrecedenceTest : AbstractBotTest
{
    private class TestBot : BaseBot
    {
        public TestBot() { }
        public TestBot(BotInfo botInfo, Uri serverUrl) : base(botInfo, serverUrl) { }
    }

    [SetUp]
    public new void SetUp()
    {
        SetAllEnvVarsToDefaultValues();
    }

    [Test]
    [Category("BOT")]
    [Property("ID", "TR-API-BOT-001c")]
    public void GivenEnvSet_whenPassingExplicitServerUrl_thenExplicitOverridesEnv()
    {
        // Arrange: set ENV to bogus port
        SetEnvVar(ServerUrl, "ws://127.0.0.1:65535");
        var botInfo = new BotInfo("Bot", "1.0", new System.Collections.Generic.List<string> { "A" },
            null, null, new System.Collections.Generic.List<string>(), new System.Collections.Generic.HashSet<string>(), null, null, null);
        var explicitUri = new Uri("ws://127.0.0.1:" + MockedServer.Port);

        // Act
        var bot = new TestBot(botInfo, explicitUri);
        StartAsync(bot);

        // Assert: connects using explicit URI
        Assert.That(Server.AwaitConnection(10_000), Is.True);
    }
}

File: bot-api/java/src/test/java/dev/robocode/tankroyale/botapi/BaseBotConstructorPrecedenceTest.java
package dev.robocode.tankroyale.botapi;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import test_utils.MockedServer;
import uk.org.webcompere.systemstubs.environment.EnvironmentVariables;
import uk.org.webcompere.systemstubs.jupiter.SystemStub;
import uk.org.webcompere.systemstubs.jupiter.SystemStubsExtension;

import java.net.URI;

import static org.assertj.core.api.Assertions.assertThat;
import static test_utils.EnvironmentVariables.*;

/**
 * TR-API-BOT-001c Precedence tests for Java: explicit args > Java System properties > ENV.
 */
import org.junit.jupiter.api.Disabled;

@ExtendWith(SystemStubsExtension.class)
@Disabled("Superseded by precedence test integrated into BaseBotConstructorTest; disabling to avoid server conflicts")
class BaseBotConstructorPrecedenceTest extends AbstractBotTest {

    static class TestBot extends BaseBot {
        TestBot() { super(); }
        TestBot(BotInfo botInfo, URI serverUrl) { super(botInfo, serverUrl); }
    }

    @SystemStub
    final EnvironmentVariables envVars = new EnvironmentVariables(
            SERVER_URL, "ws://127.0.0.1:65535", // bogus
            BOT_NAME, "TestBot",
            BOT_VERSION, "1.0",
            BOT_AUTHORS, "A,B"
    );

    @AfterEach
    void tearDown() {
        // Clear any system properties possibly set by a test
        System.clearProperty("SERVER_URL");
    }

    @Test
    void test_TR_API_BOT_001c_system_property_over_env_for_server_url() throws Exception {
        // Arrange: ENV points to a bogus port; system property points to the mocked server
        System.setProperty("SERVER_URL", "ws://127.0.0.1:" + MockedServer.PORT);

        // Act
        var bot = new TestBot();
        startAsync(bot);

        // Assert: should connect using system property URL (not ENV)
        assertThat(server.awaitConnection(10000)).isTrue();
    }

}
